File: WordPattern-Problem3.py



# dictionary approach 
        # Time complexity: O(n)
        # Space complexity: O(n)
class Solution:
    def wordPattern(self, pattern: str, str: str) -> bool:
        
        #Splitting the given sentence into words
        splitstr = str.split(" ")

        #Edge Case
        if  len(splitstr) != len(pattern):
            return False
        
        dict_res = {} # key is the pattern, value is the word
        #Logic
        for i in range(len(splitstr)):
            if pattern[i] not in dict_res:
                
                if splitstr[i] not in dict_res.values():
                    #Mapping pattern char to each word (key-pattern, value-word) 
                    dict_res[pattern[i]] = splitstr[i]
                else:
                    return False
            #if there is pattern and that patter value not equal to current word
            else:
                if  dict_res[pattern[i]] !=splitstr[i]:
                    return False
        return True

File: Word pattern-Problem 3.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution {
    public boolean wordPattern(String pattern, String s) {
         HashMap<Character,String> map1=new HashMap<>();
        HashMap<String,Character> map2=new HashMap<>();
        String[] words=s.split(" ");
        if(pattern.length()!=words.length)return false;
        for(int i=0;i<pattern.length();i++){
            if(map1.containsKey(pattern.charAt(i))){
             if(!map1.get(pattern.charAt(i)).equals(words[i])){
               return false;
             }
            }else{
                 map1.put(pattern.charAt(i),words[i]);
            }
            
             if(map2.containsKey(words[i])){
             if(!map2.get(words[i]).equals(pattern.charAt(i))){
               return false;
             }
            }else{
                 map2.put(words[i],pattern.charAt(i));
            }
        }
         
            
        return true;
    }
}